resolver/resolver.cc

```diff
@@ -1628,17 +1628,47 @@ class ResolveSignaturesWalk {
             method.data(ctx)->setReBind(sig.bind);
         }
 
+        vector<ast::Local const *> defParams; // Parameters order from the method declaration
+        bool seenOptional = false;
+
+        auto methodInfo = method.data(ctx);
+        methodInfo->resultType = sig.returns;
+
+        // Is this a signature for forward parameters?
+        if (methodInfo->arguments().size() == 3) {
+            // To match, the definition must have been desugared in exaclty 3 parameters named
+            // `<args>`, `<kargs>` and `<blockArg>`
+            auto l1 = getArgLocal(ctx, methodInfo->arguments()[0], mdef, 0, isOverloaded)->localVariable;
+            auto l2 = getArgLocal(ctx, methodInfo->arguments()[1], mdef, 1, isOverloaded)->localVariable;
+            auto l3 = getArgLocal(ctx, methodInfo->arguments()[2], mdef, 2, isOverloaded)->localVariable;
+            if (l1._name == core::Names::args() && l2._name == core::Names::kargs() &&
+                l3._name == core::Names::block()) {
+                // To provide a signature the user must specify the 3 parameters: `args`, `kargs` and `block`
+                if (sig.argTypes.size() != 3 || sig.argTypes[0].name != core::Names::args() ||
+                    sig.argTypes[1].name != core::Names::kargs() || sig.argTypes[2].name != core::Names::block()) {
+                    if (auto e = ctx.beginError(exprLoc, core::errors::Resolver::InvalidMethodSignature)) {
+                        e.setHeader("Malformed `{}` for argument forwarding. "
+                                    "Signature must declare exactly `{}` parameters named `{}`, `{}` and `{}`",
+                                    "sig", "3", core::Names::args().show(ctx), core::Names::kargs().show(ctx),
+                                    core::Names::block().show(ctx));
+                        e.addErrorLine(methodInfo->loc(), "Method declares argument forwarding here");
+                    }
+                    return;
+                }
+                // If the signature is correctly written, we replace the signatures names
+                // by the desugared parameters names
+                sig.argTypes[0].name = core::Names::args();
+                sig.argTypes[1].name = core::Names::kargs();
+                sig.argTypes[2].name = core::Names::block();
+            }
+        }
```

Comment:
After chatting with the team briefly, do you think we can punt on what syntax to use for writing signatures for these methods?

For one thing, the names of the arguments are confusing ("where did these names come from?"). If we're asking them to write the sig that way, we may as well ask them to write the method def with an explicit `def foo(*args, **kwargs, &blk)`, and then they can choose their own arg names. Consider also: it sounds like the Ruby team is considering introducing syntax like `def foo(arg0, arg1, ...)`. What would happen if users wanted to name their first arguments like `def foo(args, ...)`? This solution of choosing our own names makes it harder to support that in the future.

On top of that, chances are that the reason why people are using argument forwarding is not because they actually want to accept all keyword arguments, but it’s that they want to accept only the keyword arguments that `target(...)` accepts, and we don't have a way to write such a signature right now.

Then there's the problem that this PR doesn't implement the runtime support necessary to even get the proposed implementation of these `sig`'s working, and it's not clear what it should do there, because internally Ruby models `...` differently from `*args, **kwargs, &blk`:

```
❯ cat forward.rb
def foo(...); end

p method(:foo).arity
p method(:foo).parameters

❯ ruby forward.rb
-1
[[:rest, :*], [:block, :&]]
```

So given that there's a workaround for writing sigs for these methods (rewrite them) and the real thing that people are likely to want doesn't exist, we'd rather punt on this feature for the time being.